src/tribler/core/components/user_activity/user_activity_component.py

```diff
@@ -0,0 +1,110 @@
+from tribler.core.components.libtorrent.libtorrent_component import LibtorrentComponent
+from tribler.core.components.torrent_checker.torrent_checker.torrent_checker import TorrentChecker
+from tribler.core.components.torrent_checker.torrent_checker_component import TorrentCheckerComponent
+from tribler.core.components.user_activity.settings import UserActivitySettings
+from tribler.core.components.user_activity.types import InfoHash
+from tribler.core.sentry_reporter.sentry_reporter import SentryReporter
+
+
+class UserActivityComponent(Component):
+
+    def __init__(self, reporter: typing.Optional[SentryReporter] = None) -> None:
+        super().__init__(reporter)
+
+        self.infohash_to_queries: dict[InfoHash, list[str]] = defaultdict(list)
+        self.queries: OrderedDict[str, typing.Set[InfoHash]] = OrderedDict()
+        self.max_query_history = UserActivitySettings().max_query_history
+        self.database_manager = None
+        self.torrent_checker = None
+        self.task_manager = TaskManager()
+
+    async def run(self) -> None:
+        await super().run()
+
+        # Load settings
+        self.max_query_history = self.session.config.user_activity.max_query_history
+
+        # Wait for dependencies
+        await self.require_component(ContentDiscoveryComponent)  # remote_query_results notification
+        await self.require_component(LibtorrentComponent)  # torrent_finished notification
+        database_component = await self.require_component(DatabaseComponent)  # local_query_results notification
+        torrent_checker_component = await self.require_component(TorrentCheckerComponent)
+
+        self.database_manager: UserActivityLayer = database_component.db.user_activity_layer
+        self.torrent_checker: TorrentChecker = torrent_checker_component.torrent_checker
+
+        # Hook events
+        self.session.notifier.add_observer(notifications.torrent_finished, self.on_torrent_finished)
+        self.session.notifier.add_observer(notifications.remote_query_results, self.on_query_results)
+        self.session.notifier.add_observer(notifications.local_query_results, self.on_query_results)
+        self.task_manager.register_task("Check preferable", self.check_preferable,
+                                        interval=self.session.config.user_activity.health_check_interval)
+
+    def on_query_results(self, data: dict) -> None:
+        """
+        Start tracking a query and its results.
+
+        If any of the results get downloaded, we store the query (see ``on_torrent_finished``).
+        """
+        query = data.get("query")
+        if query is None:
+            return
+
+        results = {tmd.infohash for tmd in data["results"]}
+        for infohash in results:
+            self.infohash_to_queries[infohash].append(query)
+        self.queries[query] = results | self.queries.get(query, set())
+
+        if len(self.queries) > self.max_query_history:
```

Comment:
Perhaps for this purpose (to store a limited amount of data), you could use a dedicated data structure. This would make it possible to better cover it with tests and to use it in other parts of Tribler's code. As a beneficial side effect, it would be easier to understand the logic of the dedicated data structure and of the current method.

We can extend this draft:

```python
class LimitedSizeDict(OrderedDict):
    def __init__(self, *args, size_limit=None, **kwargs):
        super().__init__(*args, **kwargs)
        self.size_limit = size_limit

    def __setitem__(self, key, value):
        super().__setitem__(key, value)
        self._check_size_limit()

    def _check_size_limit(self):
        if self.size_limit is None:
            return
        while len(self) > self.size_limit:
            self.popitem(last=False)
```